elixir/word-count/word_count.exs

```diff
@@ -0,0 +1,17 @@
+defmodule Words do
+  @doc """
+  Count the number of words in the sentence.
+
+  Words are compared case-insensitively.
+  """
+  @spec count(String.t()) :: map
+  def count(sentence) do
+    Enum.reduce(
+      Regex.scan(~r/[a-zA-Z0-9-\p{L}]+/u, sentence),
```

Comment:
Well... it works, so there's not much to complain here :p
However, some things here deviate from the way I usually craft Elixir code and see it around in the wild:

## First thing I miss here is the idea of a pipeline for data transformation.
I usually try to think the code as a step-by-step guide on how to do the transformation I want. So I try to abuse the pipe `|>` operator. E.g.:
```
sentence
|> do_first_step
|> map_to_something
|> maybe_more_transformations
|> reduce_to_something
|> replace_whatever_needs_to_be_replaced
|> i_have_no_idea_what_i_am_doing_but_you_got_the_idea
```
The way it is right now it looks to me like we are trying to use one of those python one-liners that are nice and cool, but a bit hard to understand for those who are not used to it. What is this `Regex.scan/2` doing to the sentence? Could it be done with other function calls? Usually doing the step-by-step pipeline I used as an example makes things easier to me.

## Another thing is the function name
I know this is given by the exercise, but just my 2 cents. You might have felt the same way. I just want to promote some discussion :p
`count` and returning a `map` looks odd. Speacially when you are getting a `String.t()` as a parameter.
What are you counting here? Why does count return a map? Can you think of other function names that might be better to represent what we are trying to accomplish?

Reply:
Wow! Thanks for the review 💜 

About the pipelines, I understand your feeling 💜 
I know that it is easier to follow (because it is a pattern and it is also more explicit)

But I couldn't find a way to start from `sentence` and add the `Regex.scan` in the pipeline without adding more methods, because the second method expects the sentence as the second arg instead of first (@msramos resolved it 💜).
I know that to add more methods is ok but, considering the scope of this exercise and the goal of learning elixir while solving challenges, in my vision, it could increase the indirection without improving significantly the readability.

Your wall of text brought me good reflections 💜 

> count and returning a map looks odd
Actually, I didn't think about it 😆 
But I don't know which name could be better in this case 🤔 
Something like `count_words`

P.S.:  I'm feeling now like when I learned how to drive a car
I have to pay attention to so so many things 😱

Reply:
Using pipes is like telling a story. Using your code:

> I have this `@pattern` that I want to use to `scan` a `sentence` to find individual words. Then I want to count the frequency of each word.

The last part, though, isn't so obvious by just looking at the code. We can extract it to hide the details from the pipeline:

```elixir
@pattern ~r/[a-zA-Z0-9-\p{L}]+/u

def count(sentence) do
  @pattern
  |> Regex.scan(sentence)
  |> count_frequency()
end

defp count_frequency(list) do
  Enum.reduce(list, %{}, fn [word], acc ->
    Map.update(acc, String.downcase(word), 1, &(&1 + 1))
  end
end
```

This is just an example, of course. Maybe it's too overkill for this exercise :)

Reply:
`s/method/function` 👀

Following this idea of having the pipeline, I would say that the transformation `String.downcase(word)` and even the pattern matching `[word]` on the reduction could be made more explicit. A suggestion, quibeing msramos' a little bit, could be:

```elixir
def count(sentence) do
  @word_pattern
  |> Regex.scan(sentence)
  |> List.flatten()
  |> Enum.map(&String.downcase/1)
  |> Enum.reduce(%{}, fn word, acc ->
    Map.update(acc, word, 1, &(&1 + 1))
  end)
end
```

I'd rather have more explicit names (creating functions and/or variables), but what you said makes sense. 🙂

Reply:
@Lucasnar while more readable, it's worth mentioning that this code is traversing the the list (returned by `Regex.scan/2) three times: flatten, map and reduce. The original code does everything in just one reduce.

Reply:
As @lucasnar pointed, we call everything functions 👀 

The whole sentence could be down-cased before running the regex to tackle the problem @msramos brought. That way we also get a more readable reduce function, since the original one has quite an amount of nested function calls.

Personally, I solved this problem without Regex, but using more transformations on the original sentence 😬 

Think about the story example:
```
- We start with a `pattern`;
- We scan over some `sentence`;
- We generate a map with what returned from the previous step;
```
That is a nice and functional story, but I feel the protagonist is the `pattern`, not `sentence`...
I liked our little fella `sentence` :(

So my story goes like this:
```
- We start with a sentence;
- We ignore underscores;
- We make it case-insensitive (down-casing);
- We split it into words;
- We count the words;
```

Try to come up with the code if you guys want and see if it looks more readable or not.
I liked it better the way I pointed (ohh the bias :P), although I have to say using regex looks more "I know what I am doing here".

Reply:
Yeah, but since this is a micro-optimization, I would choose readability over this performance gain. When running the tests here, there was virtually no difference in the test results (sometimes my solution even had better results haha). I agree it is important to keep that in mind, though.